Port to python3, dockerize and add some args

simeonmiteff opened the pull request:
Hi @thepacketgeek 

Thanks for this tool. I've:
- Made it work with @ofalk's (fork?) of libdnet: http://github.com/ofalk/libdnet
- Converted to Python 3
- Added a Dockerfile
- Added an option to overwrite instead of append to output files (now default)`*`
- Added a VLAN tag rewriting feature (fixed VLAN only, no support for multiple tags and no randomisation)

`*` not sure if this is wise in general but is _my_ preference.

Commit:
Add option to override vlan tag

Commit:
Fix python2->3 issue: open as binary file

Commit:
Convert with 2to3

Commit:
Make buildable in ubuntu:focal Docker image

Commit:
Add arg to select append vs overwrite (make default overwrite)

Commit:
Add bangline

simeonmiteff commented:
Oh I forgot to add - I haven't changed the documentation (README.md) - I'd like to hear where you'd like to go with this PR before reworking it.

thepacketgeek commented:
Thank you for bring a fresh breath of air to this project :) I'll look through the commits and let you know, I think all of these are wonderful changes bringing sanicap up to date!

thepacketgeek commented on sanicap/sanicap.py:
```diff
@@ -51,7 +53,7 @@ def _random_mac(self, address):
             return format(int('0x' + unpadded.replace(':','').replace('.',''), 16), '048b')
 
-        unmasked = ''.join([str(randint(0,1)) for x in xrange(0, 48 - self.mask)])
+        unmasked = ''.join([str(randint(0,1)) for x in range(0, 48 - self.mask)])
```
I realize this was existing: join works with an iterator, no need to create an intermediate list
```suggestion
        unmasked = ''.join(str(randint(0,1)) for x in range(0, 48 - self.mask))
```

thepacketgeek commented on sanicap/sanicap.py:
```diff
@@ -113,9 +115,9 @@ def pad_bin(unpadded):
+            return format(int(ipaddress.IPv4Address(str(unpadded))), '032b')
 
-        unmasked = ''.join([str(randint(0,1)) for x in xrange(0, 32 - self.mask)])
+        unmasked = ''.join([str(randint(0,1)) for x in range(0, 32 - self.mask)])
```
Same as above
```suggestion
        unmasked = ''.join(str(randint(0,1)) for x in range(0, 32 - self.mask))
```

thepacketgeek commented on sanicap/sanicap.py:
```diff
@@ -181,9 +183,9 @@ def pad_bin(unpadded):
+            return format(int(ipaddress.IPv6Address(str(unpadded))), '0128b')
 
-        unmasked = ''.join([str(randint(0,1)) for x in xrange(0, 128 - self.mask)])
+        unmasked = ''.join([str(randint(0,1)) for x in range(0, 128 - self.mask)])
```
Same as above
```suggestion
        unmasked = ''.join(str(randint(0,1)) for x in range(0, 128 - self.mask))
```

thepacketgeek commented on sanicap/sanicap.py:
```diff
@@ -217,7 +219,7 @@ def get_ip(self, address):
 
 
-def sanitize(filepath_in, filepath_out = None, sequential=True, ipv4_mask=0, ipv6_mask=0, mac_mask=0, start_ipv4='10.0.0.1', start_ipv6='2001:aa::1', start_mac='00:aa:00:00:00:01', info=True):
+def sanitize(filepath_in, filepath_out = None, sequential=True, append=False, ipv4_mask=0, ipv6_mask=0, mac_mask=0, start_ipv4='10.0.0.1', start_ipv6='2001:aa::1', start_mac='00:aa:00:00:00:01', fixed_vlan=None, info=True):
```
In lieu of an `Options` or `Config` class for all of these options, making them kwarg only should help usage stay readable.

```suggestion
def sanitize(filepath_in, *, filepath_out = None, sequential=True, append=False, ipv4_mask=0, ipv6_mask=0, mac_mask=0, start_ipv4='10.0.0.1', start_ipv6='2001:aa::1', start_mac='00:aa:00:00:00:01', fixed_vlan=None, info=True):
```

thepacketgeek reviewed (changes requested):
Love it! Just a few small changes, thank you for this

simeonmiteff committed:
Join iterator without intermediate list

Co-authored-by: Mat Wood <thepacketgeek@users.noreply.github.com>

simeonmiteff committed:
Make sanitize() arguments [1:] kwarg-only

Co-authored-by: Mat Wood <thepacketgeek@users.noreply.github.com>

thepacketgeek commented:
This is great, thanks again! Let me know if you are still planning on README changes or if this is ready to merge

Commit:
Improve error handling

- Print e, not e.message
- Exit with nonzero status
- Improve error message when opening pcapng files

Commit:
Optimise imports

Commit:
PEP8 reformat

Commit:
Make unused range variables _

Commit:
Rework README.md

simeonmiteff commented:
@thepacketgeek I just realised sanicap doesn't need `dnet` :facepalm:

simeonmiteff commented:
> This is great, thanks again! Let me know if you are still planning on README changes or if this is ready to merge

I reworked the README and while generating an example output I discovered that the kwarg change introduced a bug, so I fixed that and amended that commit. I also fixed/improved the error handling and fixed some formatting things that PyCharm complained about.

I also revised the Dockerfile commit and ripped all the `libdnet` building garbage.

The Dockerfile is specifically useful to be because I'm running sanicap on hosts without web access so just scp-ing the image over is easier than sorting out offline installs of the 3 dependencies. I'd like to keep it in the repo but I recon this module really also needs proper python packaging. I'm reaching the limit of how much time I can justify hacking on this now.

TBH I'd love to write something like sanicap in Rust (not just for RIIR sake: sanicap is _slow_ on a 1.5GB PCAP). I have some good code for handling both pcap and pcapng inputs but scapy does a huge amount of heavy lifting dealing with the link layer all the way up the stack - that would be hard to replicate in Rust (unless there are some awesome crates I have yet to discover).

thepacketgeek commented:
The Dockerfile does seem very useful good to keep in the repo. 

I agree that this project could benefit greatly from a rewrite in Rust, it's something that I was pondering when I realized there was no scapy equivalent in Rust. I started on some of the underlying features for the send/receive packet flows in [`rust-network-tools`](https://github.com/thepacketgeek/rust-network-tools), but modifying and crafting packets has been a pain in the libraries I've tried so far ([`etherparse`](https://crates.io/crates/resynth) and [`pnet`](https://crates.io/crates/pnet)). [`Resynth`](https://crates.io/crates/resynth) looks promising but it'll be a while before anything catches up to Scapy with protocol support.

ofalk commented:
[ ... ]
> Thanks for this tool. I've:
>     * Made it work with @ofalk's (fork?) of libdnet: [http://github.com/ofalk/libdnet](https://github.com/ofalk/libdnet)

It used to be a fork, but quite a while ago I took over the project from it's original author and now it's the official one, just FYI.

ofalk commented:
> @thepacketgeek I just realised sanicap doesn't need `dnet` facepalm

Great, I'm out then :-)
